tools/plugins/com.liferay.ide.upgrade.problems.core/src/com/liferay/ide/upgrade/problems/core/internal/tasks/FindUpgradeProblemsTaskStepAction.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.ide.upgrade.problems.core.internal.tasks;
+
+import com.liferay.ide.core.util.FileUtil;
+import com.liferay.ide.upgrade.plan.core.BaseUpgradeTaskStepAction;
+import com.liferay.ide.upgrade.plan.core.UpgradePlanner;
+import com.liferay.ide.upgrade.plan.core.UpgradeTaskStepAction;
+import com.liferay.ide.upgrade.plan.core.UpgradeTaskStepActionDoneEvent;
+import com.liferay.ide.upgrade.problems.core.FileMigration;
+import com.liferay.ide.upgrade.problems.core.FileUpgradeProblem;
+import com.liferay.ide.upgrade.problems.core.UpgradeProblemsCorePlugin;
+import com.liferay.ide.upgrade.tasks.core.ResourceSelection;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Stream;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.debug.internal.core.XMLMemento;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ServiceScope;
+
+/**
+ * @author Terry Jia
+ */
+@Component(
+	property = {"id=find_upgrade_problems", "order=1", "stepId=find_upgrade_problems", "title=Find Upgrade Problems"},
+	scope = ServiceScope.PROTOTYPE, service = UpgradeTaskStepAction.class
+)
+@SuppressWarnings("restriction")
+public class FindUpgradeProblemsTaskStepAction extends BaseUpgradeTaskStepAction {
+
+	@Override
+	public IStatus perform() {
+		List<IProject> projects = _resourceSelection.selectProjects("select projects", true);
+
+		List<String> versions = new ArrayList<>();
+
+		versions.add("7.0");
+		versions.add("7.1");
+
+		Job job = new Job("Finding migration problems...") {
+
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				XMLMemento root = XMLMemento.createWriteRoot("upgradeProblems");
```

Comment:
When the problems are found, we should just set the problems into the current upgradepPlan.

`UpgradePlan.addFileUpgradeProblems(...)`

Then we can let the `upgradePlan` class decide how best to store these prolems.  We don't want this class `FindUpgradeProblemsTaskStepAction` to know how other classes are going to store (or are expected to store the results).  

This TaskStepAction should jus fine the problems and then send them over to the upgradePlan and the plan will deal with how to store them.